### PR TITLE
Updates CUDA memory leak check to verify against driver AP, print more diagnostic information, and perform fewer garbage collections

### PR DESCRIPTION
**Pull request by mruberry**
Per title

### CODE REVIEW

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -1210,35 +1213,110 @@ def __init__(self, testcase, name=None):
...
+                break
+
+        # Short-circuits if no discrepancy detected
+        if not discrepancy_detected:
+            return
+
+        # See #62533
+        # ROCM: Sometimes the transient memory is reported as leaked memory
+        if TEST_WITH_ROCM:
+            return
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1267]: Add warning back

**Review by mruberry — COMMENTED** [review]
TODO:
- add warning back
- fix lint
- revert environment variable change (after letting current CI run to completion)

**Review by ngimel — APPROVED** [review]

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -770,8 +770,11 @@ def _check_module_exists(name: str) -> bool:
 # Determine whether to enable cuda memory leak check.
 # CUDA mem leak check is expensive and thus we don't want to execute it on every
 # test case / configuration.
+# If this is True then CUDA memory leak checks are skipped. If this is false
+#   then CUDA memory leak checks are performed.
 # See: https://github.com/pytorch/pytorch/pull/59402#issuecomment-858811135
-TEST_SKIP_CUDA_MEM_LEAK_CHECK = os.getenv('PYTORCH_TEST_SKIP_CUDA_MEM_LEAK_CHECK', '0') == '1'
+# TEST_SKIP_CUDA_MEM_LEAK_CHECK = os.getenv('PYTORCH_TEST_SKIP_CUDA_MEM_LEAK_CHECK', '0') == '1'
+TEST_SKIP_CUDA_MEM_LEAK_CHECK = False  # TODO: update me!
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:777]: Be sure to revert this

File: mypy.ini [hunk 3]
```diff
@@ -41,7 +41,7 @@ files =
 #
 # `exclude` is a regex, not a list of paths like `files` (sigh)
 #
-exclude = torch/include/|torch/csrc/|torch/distributed/elastic/agent/server/api.py
+exclude = torch/include/|torch/csrc/|torch/distributed/elastic/agent/server/api.py|torch/testing/_internal
```
Comment (malfet) [inline mypy.ini:44]: Just curious - what's the point of removing torch/testing/_internal from mypy checks?

Comment (mruberry) [reply mypy.ini:44]: I'll create a post

**Pull request closed without merging**